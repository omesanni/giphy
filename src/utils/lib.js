import React from 'react';
import numeral from 'numeral';
import AsyncComponent from '../components/AsyncComponent';

/**
 * Check if a number is within range for pagination, etc.
 * @param {Object} range
 * @param {Number} i
 * @param {Number} total
 */
export function isWithinRange(range, i, total) {
  return (
    i <= range.max &&
    i >= range.min &&
    i <= total &&
    i > 0
  );
}

/**
 * Set a range for pagination, etc.
 * @param {Number} current
 * @param {Number} spread
 * @param {Number} total
 */
export function setRange(current, spread, total) {
  if (current <= (1 + spread)) {
    return {
      min: 1,
      max: Math.min(1 + (spread * 2), total),
    };
  } else if (current >= (total - spread)) {
    return {
      min: Math.max(total - (spread * 2), 1),
      max: total,
    };
  }

  return {
    min: current - spread,
    max: Math.min(current + spread, total),
  };
}

/**
 * Formats numbers
 * @param {Number} number
 * @param {String} format
 * @return {String}
 */
export function formatNumber(number = 0, format = '0,0[.]00') {
  return numeral(number).format(format);
}

/**
 * Store data in local storage
 * @param  {String} key
 * @param  {Object} data
 * @return {String}
 */
export function saveToLocalStorage(key, data) {
  return localStorage.setItem(key, JSON.stringify(data));
}

/**
 * Fetch data from local storage
 * @param  {String} key
 * @return {Object}
 */
export function fetchFromLocalStorage(key) {
  return JSON.parse(localStorage.getItem(key));
}

/**
 * Converts object to query string
 * @param  {Object} query Query object
 * @return {String} queryStr Query string
 */
export function stringifyQuery(query = {}) {
  let queryStr = '';

  const keys = Object.keys(query);
  const lastIndex = keys.length - 1;

  keys.forEach((key, i) => {
    queryStr += `${key}=${query[key]}${(lastIndex !== i) ? '&' : ''}`;
  });

  return queryStr;
}

/**
 * Lazy loads a component
 * @param  {Function} getComponent Returns a dynamic import of the component
 * @param  {Object}   props
 * @return {ReactElement}
 */
export function lazyLoadComponent(getComponent, props = {}) {
  return React.createElement(AsyncComponent, { ...props, render: getComponent });
}
