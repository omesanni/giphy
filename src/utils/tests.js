import renderer from 'react-test-renderer';

/**
 * Checks if component matches snapshot created by jest
 * @param  {ReactElement} component
 */
export function expectToMatchSnapshot(component) {
  expect(renderer.create(component).toJSON()).toMatchSnapshot();
}
