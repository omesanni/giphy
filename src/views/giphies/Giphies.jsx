import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import isEqual from 'lodash/isEqual';
import numeral from 'numeral';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faSearch from '@fortawesome/fontawesome-free-solid/faSearch';
import { formatNumber, saveToLocalStorage } from '../../utils/lib';
import Loader from '../../components/Loader';
import Modal from '../../components/Modal';
import RankControls from '../../components/RankControls';
import Pagination from '../../components/Pagination';
import * as giphyActions from './actions';
import * as modalActions from '../../shared/modals/actions';

export class Giphies extends React.Component {
  constructor() {
    super();

    this.modalId = 'giphy-details';
    this.searchNode = null;
    this.pageLimitOptions = [10, 25, 50];
    this.sortingOptions = ['asc', 'desc'];

    this.state = {
      pageLimit: 25,
      searchVal: '',
      focusedGiphy: null,
      sortDirection: null,
    };

    this.handleSearch = this.handleSearch.bind(this);
    this.handleRanking = this.handleRanking.bind(this);
    this.renderGiphyCards = this.renderGiphyCards.bind(this);
    this.renderHeaderControls = this.renderHeaderControls.bind(this);
    this.renderGiphyDetailsModal = this.renderGiphyDetailsModal.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    const { focusedGiphy } = this.state;

    if (focusedGiphy) {
      const { records: nextRecords } = nextProps.giphies;
      const { records } = this.props.giphies;

      const nextRecord = nextRecords.find(n => (n.id === focusedGiphy.id));
      const record = records.find(g => (g.id === focusedGiphy.id));

      if (!isEqual(record, nextRecord)) {
        this.setState(() => ({ focusedGiphy: nextRecord }));
      }
    }
  }

  /**
   * Handler for performing giphy search
   * @param  {Number}  offset           Page offset
   * @param  {Boolean} paginationFetch  Flag to detect when page number is clicked
   * @return {Promise}
   */
  handleSearch(offset = 0, paginationFetch = false) {
    const { actions } = this.props;
    const { value } = this.searchNode;
    const { sortDirection, pageLimit, searchVal } = this.state;

    // use value saved in state if a page number is clicked else use
    // search input value
    const preferredVal = paginationFetch ? searchVal : value;

    if (!preferredVal) {
      return;
    }

    // query object in the format needed by giphy API
    const query = {
      q: preferredVal,
      limit: pageLimit,
      offset,
    };

    return actions.fetchGiphies(query)
      .then((res) => {
        this.setState(() => ({ searchVal: preferredVal }));

        // sort records after
        if (res.data && sortDirection) {
          actions.sortGiphies(sortDirection);
        }
      });
  }

  /**
   * Upvote or downvote giphy
   * @param  {Object}  giphy    Particular giphy record
   * @param  {Boolean} upvoting Flag for upvoting or downvoting
   */
  handleRanking(giphy, upvoting = true) {
    const FRESH_SLATE = !giphy.upvoted && !giphy.downvoted;

    const UPVOTED_BEFORE_AND_UPVOTING = giphy.upvoted && upvoting;
    const UPVOTED_BEFORE_AND_DOWNVOTING = giphy.upvoted && !upvoting;

    const DOWNVOTED_BEFORE_AND_DOWNVOTING = giphy.downvoted && !upvoting;
    const DOWNVOTED_BEFORE_AND_UPVOTING = giphy.downvoted && upvoting;

    let downvoted, upvoted;
    let score = giphy._score;

    switch (true) {
      case UPVOTED_BEFORE_AND_UPVOTING || UPVOTED_BEFORE_AND_DOWNVOTING: {
        score -= 1;
        upvoted = !giphy.upvoted;
        downvoted = UPVOTED_BEFORE_AND_DOWNVOTING;
        break;
      }
      case DOWNVOTED_BEFORE_AND_DOWNVOTING || DOWNVOTED_BEFORE_AND_UPVOTING: {
        score += 1;
        downvoted = !giphy.downvoted;
        upvoted = DOWNVOTED_BEFORE_AND_UPVOTING;
        break;
      }
      case FRESH_SLATE && upvoting: {
        score += 1;
        upvoted = true;
        downvoted = !!downvoted;
        break;
      }
      case FRESH_SLATE && !upvoting: {
        score -= 1;
        downvoted = true;
        upvoted = !!upvoted;
        break;
      }
    }

    const update = {
      ...giphy,
      _score: score,
      upvoted,
      downvoted,
    };

    this.props.actions.updateGiphy(update);

    // save to local storage
    saveToLocalStorage(update.id, update);
  }

  /**
   * Renders Modal with giphy details
   * @return {JSX}
   */
  renderGiphyDetailsModal() {
    const { modals, actions } = this.props;
    const { focusedGiphy } = this.state;

    return (
      <Modal
        id={this.modalId}
        actions={actions}
        modals={modals}
      >
        {focusedGiphy && (
          <div className={'row'}>
            <div className={'col-sm-6 col-md-5'}>
              <img
                className={'card-img h-75'}
                src={focusedGiphy.images.downsized_medium.url}
              />

              <RankControls
                score={formatNumber(focusedGiphy._score)}
                onUpvote={() => this.handleRanking(focusedGiphy)}
                onDownvote={() => this.handleRanking(focusedGiphy, false)}
                className={'mt-3'}
              />
            </div>

            <div className={'col-sm-6 col-md-7'}>
              <ul className={'list-group'}>
                {focusedGiphy.title && (
                  <li className={'list-group-item d-flex'}>
                    <strong className={'mr-1'}>
                      {'Title:'}
                    </strong>
                    <span
                      title={focusedGiphy.title}
                      className={'text-capitalize text-truncate'}
                    >
                      {focusedGiphy.title}
                    </span>
                  </li>
                )}
                {focusedGiphy.username && (
                  <li className={'list-group-item d-flex'}>
                    <strong className={'mr-1'}>
                      {'Username:'}
                    </strong>
                    <span>
                      {focusedGiphy.username}
                    </span>
                  </li>
                )}
                <li className={'list-group-item d-flex'}>
                  <strong className={'mr-1'}>
                    {'Rating:'}
                  </strong>
                  <span className={'text-uppercase'}>
                    {focusedGiphy.rating}
                  </span>
                </li>
                <li className={'list-group-item d-flex'}>
                  <strong className={'mr-1'}>
                    {'Embed:'}
                  </strong>
                  <a
                    href={focusedGiphy.embed_url}
                    title={focusedGiphy.embed_url}
                    target={'_blank'}
                    className={'text-truncate'}
                  >
                    {focusedGiphy.embed_url}
                  </a>
                </li>
                {focusedGiphy.source && (
                  <li className={'list-group-item d-flex'}>
                    <strong className={'mr-1'}>
                      {'Source:'}
                    </strong>
                    <a
                      href={focusedGiphy.source}
                      title={focusedGiphy.source}
                      target={'_blank'}
                      className={'text-truncate'}
                    >
                      {focusedGiphy.source}
                    </a>
                  </li>
                )}
                <li className={'list-group-item d-flex'}>
                  <strong className={'mr-1'}>
                    {'Score:'}
                  </strong>
                  <span>
                    {formatNumber(focusedGiphy._score)}
                  </span>
                </li>
              </ul>
            </div>
          </div>
        )}
      </Modal>
    );
  }

  /**
   * Renders search input and page limit controls
   * @return {JSX}
   */
  renderHeaderControls() {
    return (
      <div className={'form-row'}>
        <div className={'form-group col-lg-4 col-sm-6 col-md-5 mb-0 mr-auto'}>
          <div className={'input-group'}>
            <input
              id={'search-input'}
              type={'text'}
              ref={node => (this.searchNode = node)}
              placeholder={'Search...'}
              className={'form-control form-control-sm'}
            />

            <div className={'input-group-append'}>
              <button
                id={'search-btn'}
                type={'button'}
                onClick={() => this.handleSearch()}
                className={'btn btn-outline-secondary'}
              >
                <FontAwesomeIcon icon={faSearch} />
              </button>
            </div>
          </div>
        </div>

        <div className={'form-group col-lg-2 col-sm-3 justify-content-end d-flex mb-0'}>
          <label className={'mr-3 d-none d-sm-block text-muted'}>
            {'Limit'}
          </label>
          <select
            id={'page-limit-select'}
            className={'form-control form-control-sm'}
            defaultValue={this.state.pageLimit}
            onChange={(e) => {
              const { value } = e.target;

              this.setState(() => ({
                pageLimit: value,
              }), () => this.handleSearch(0, true));
            }}
          >
            {this.pageLimitOptions.map((opt) => (
              <option key={opt}>{opt}</option>
            ))}
          </select>
        </div>

        <div className={'form-group col-lg-1 col-sm-2 mb-0'}>
          <select
            id={'sort-select'}
            className={'form-control form-control-sm'}
            defaultValue={'Order By'}
            onChange={(e) => {
              const { value } = e.target;

              this.setState(() => ({
                sortDirection: value,
              }), () => this.props.actions.sortGiphies(value));
            }}
          >
            <option disabled>{'Order By'}</option>
            {this.sortingOptions.map((dir) => (
              <option key={dir}>{dir}</option>
            ))}
          </select>
        </div>
      </div>
    );
  }

  /**
   * Renders giphy results
   * @return {JSX}
   */
  renderGiphyCards() {
    const { records, fetching, error, pagination } = this.props.giphies;
    const { pages, current } = pagination;

    return (
      <div className={'card-body'}>
        <div className={'position-relative'}>
          {fetching && <Loader />}

          <div className={'row'}>
            {records.map((giphy) => (
              <div
                key={giphy.id}
                className={'col-sm-6 col-md-4 col-lg-3 mt-3'}
              >
                <div
                  id={`giphy-${giphy.id}`}
                  className={'card'}
                >
                  <img
                    className={'object-fit-cover cursor-pointer rounded-top'}
                    src={giphy.images.fixed_height_downsampled.url}
                    alt={giphy.title}
                    onClick={() => {
                      this.props.actions.openModal(this.modalId);
                      this.setState(() => ({ focusedGiphy: giphy }));
                    }}
                  />

                  <div className={'card-body card-body--rank-wrapper'}>
                    <RankControls
                      score={formatNumber(giphy._score)}
                      onUpvote={() => this.handleRanking(giphy)}
                      onDownvote={() => this.handleRanking(giphy, false)}
                    />
                  </div>
                </div>
              </div>
            ))}

            {(error || !records.length) && (
              <div
                id={error ? 'network-error' : 'no-giphies'}
                className={'col d-flex justify-content-center text-muted'}
              >
                <h5 className={'font-weight-light'}>
                  {error ? 'Network Error' : 'No Giphies to display'}
                </h5>
              </div>
            )}
          </div>

          {!!records.length && (
            <Pagination
              current={current}
              pages={pages}
              className={'mt-5'}
              onChange={(page) => {
                const offset = (page - 1) * this.state.pageLimit;

                this.handleSearch(offset, true);
              }}
            />
          )}
        </div>
      </div>
    );
  }

  render() {
    return (
      <section className={'card card--emboss mt-5'}>
        <div className={'card-header'}>
          {this.renderHeaderControls()}
        </div>

        {this.renderGiphyCards()}
        {this.renderGiphyDetailsModal()}
      </section>
    );
  }
}

Giphies.propTypes = {
  actions: PropTypes.objectOf(PropTypes.func).isRequired,
  modals: PropTypes.objectOf(PropTypes.object).isRequired,
  giphies: PropTypes.shape({
    records: PropTypes.array,
    error: PropTypes.bool,
    fetching: PropTypes.bool,
    pagination: PropTypes.object,
  }).isRequired,
};

export default connect(state => ({
  modals: state.get('modals').toJS(),
  giphies: state.get('giphies').toJS(),
}), dispatch => ({
  actions: bindActionCreators({
    ...giphyActions,
    ...modalActions,
  }, dispatch),
}))(Giphies);
