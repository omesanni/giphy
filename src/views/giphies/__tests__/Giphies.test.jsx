import React from 'react';
import { shallow } from 'enzyme';
import { Giphies } from '../Giphies';
import { expectToMatchSnapshot } from '../../../utils/tests';
import { fetchFromLocalStorage } from '../../../utils/lib';

describe('Giphies', () => {
  let component;

  const records = [
    {
      id: 'aaa',
      _score: 20,
      title: 'cash me outside how bout dah',
      username: 'dirtydozen',
      rating: 'G',
      source: 'http://youtube',
      embed_url: 'http://embed',
      images: {
        downsized_medium: { url: 'http//www' },
        fixed_height_downsampled: { url: 'http//www' },
      },
    },
    {
      id: 'bbb',
      _score: 30,
      title: 'put some respeck on my name',
      username: 'southside',
      rating: 'PG',
      source: 'http://youtube',
      embed_url: 'http://embed',
      images: {
        downsized_medium: { url: 'http//www' },
        fixed_height_downsampled: { url: 'http//www' },
      },
    },
    {
      id: 'ccc',
      _score: 10,
      rating: 'PG',
      embed_url: 'http://embed',
      images: {
        downsized_medium: { url: 'http//www' },
        fixed_height_downsampled: { url: 'http//www' },
      },
    },
  ];

  const giphies = {
    error: false,
    fetching: false,
    pagination: { current: 1, pages: records.length },
    records,
  };

  const props = {
    modals: {
      'giphy-details': { isOpen: true },
    },
    actions: {
      openModal: jest.fn(),
      sortGiphies: jest.fn(),
      createModalStore: jest.fn(),
      deleteModalStore: jest.fn(),
      fetchGiphies: jest.fn(() => Promise.resolve({ data: {} })),
      updateGiphy: jest.fn((update) => {
        const newGiphies = {
          ...giphies,
          records: records.map((record) => {
            return record.id === update.id ? update : record;
          }),
        };

        component.setProps({ giphies: newGiphies });
      }),
    },
    giphies,
  };

  function getGiphies(p = props) {
    return <Giphies {...p} />;
  }

  function mountComponent(ownProps = props) {
    return shallow(getGiphies(props));
  }

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders', () => {
    expectToMatchSnapshot(getGiphies());
  });

  it('should sort giphies', () => {
    component = mountComponent();

    component.find('#sort-select').simulate('change', {
      target: { value: 'desc' },
    });

    expect(props.actions.sortGiphies).toHaveBeenCalledWith('desc');
  });

  it('should show loader animation while fetching giphies', () => {
    component = mountComponent();

    component.setProps({ giphies: { ...props.giphies, fetching: true } });

    expect(component.find('Loader').length).toBe(1);
  });

  it('should show network error message', () => {
    component = mountComponent();

    component.setProps({ giphies: { ...props.giphies, error: true } });

    expect(component.find('#network-error').length).toBe(1);
  });

  it('should show no giphies message when no records were returned', () => {
    component = mountComponent();

    component.setProps({ giphies: { ...props.giphies, records: [] } });

    expect(component.find('#no-giphies').length).toBe(1);
  });

  it('should upvote giphy', () => {
    component = mountComponent();
    const instance = component.instance();

    jest.spyOn(instance, 'handleRanking');
    component.find(`#giphy-${records[0].id}`).find('RankControls').props().onUpvote();

    expect(instance.handleRanking).toHaveBeenCalledWith(records[0]);
  });

  it('should downvote giphy', () => {
    component = mountComponent();
    const instance = component.instance();

    jest.spyOn(instance, 'handleRanking');
    component.find(`#giphy-${records[0].id}`).find('RankControls').props().onDownvote();

    expect(instance.handleRanking).toHaveBeenCalledWith(records[0], false);
  });

  describe('openModal', () => {
    it('should open modal and show extra details when giphy is clicked', () => {
      component = mountComponent();
      const instance = component.instance();

      component.find(`#giphy-${records[2].id}`).find('img').simulate('click');

      expect(instance.state.focusedGiphy).toEqual(records[2]);
      expect(props.actions.openModal).toHaveBeenCalledWith(instance.modalId);
    });

    it('should update `focusedGiphy` state when giphy is upvoted or downvoted', () => {
      component = mountComponent();
      component.find(`#giphy-${records[1].id}`).find('img').simulate('click');

      const oldState = component.instance().state.focusedGiphy;
      component.find('Modal').find('RankControls').props().onUpvote();

      const newState = component.instance().state.focusedGiphy;

      expect(oldState).not.toEqual(newState);
    });
  });

  describe('handleRanking()', () => {
    it('should increment score and set `upvoted` to true', () => {
      component = mountComponent();
      const giphy = { id: 'aba', _score: 5 };

      component.instance().handleRanking(giphy);
      const update = fetchFromLocalStorage(giphy.id);

      expect(update._score).toEqual(giphy._score + 1);
      expect(update.upvoted).toBe(true);
    });

    it('should decrement score if someone who upvoted before is upvoting again', () => {
      component = mountComponent();
      const giphy = { id: 'bcc', _score: 8, upvoted: true };

      component.instance().handleRanking(giphy);
      const update = fetchFromLocalStorage(giphy.id);

      expect(update._score).toEqual(giphy._score - 1);
      expect(update.upvoted).toBe(false);
    });

    it('should decrement score and set `downvoted` to true', () => {
      component = mountComponent();
      const giphy = { id: 'ddc', _score: 10 };

      component.instance().handleRanking(giphy, false);
      const update = fetchFromLocalStorage(giphy.id);

      expect(update._score).toEqual(giphy._score - 1);
      expect(update.downvoted).toBe(true);
    });

    it('should increment score if someone who downvoted before is downvoting again', () => {
      component = mountComponent();
      const giphy = { id: 'xcc', _score: 20, downvoted: true };

      component.instance().handleRanking(giphy, false);
      const update = fetchFromLocalStorage(giphy.id);

      expect(update._score).toEqual(giphy._score + 1);
      expect(update.downvoted).toBe(false);
    });

    it('should decrement score if someone who upvoted before is now downvoting', () => {
      component = mountComponent();
      const giphy = { id: 'ddd', _score: 11, upvoted: true };

      component.instance().handleRanking(giphy, false);
      const update = fetchFromLocalStorage(giphy.id);

      expect(update._score).toEqual(giphy._score - 1);
      expect(update.upvoted).toBe(false);
      expect(update.downvoted).toBe(true);
    });

    it('should increment score if someone who downvoted before is now upvoting', () => {
      component = mountComponent();
      const giphy = { id: 'bbx', _score: 3, downvoted: true };

      component.instance().handleRanking(giphy);
      const update = fetchFromLocalStorage(giphy.id);

      expect(update._score).toEqual(giphy._score + 1);
      expect(update.upvoted).toBe(true);
      expect(update.downvoted).toBe(false);
    });
  });

  describe('handleSearch()', () => {
    it('should fetch giphies from API when search button is clicked', () => {
      component = mountComponent();

      component.instance().searchNode = { value: 'drake' };
      component.find('#search-btn').simulate('click');

      expect(props.actions.fetchGiphies).toHaveBeenCalled();
    });

    it('should fetch giphies from API when page limit is changed', () => {
      component = mountComponent();
      const instance = component.instance();

      instance.searchNode = {}
      instance.setState(() => ({ searchVal: 'demar defrozen' }));

      component.find('#page-limit-select').simulate('change', {
        target: { value: 50 },
      });

      expect(props.actions.fetchGiphies).toHaveBeenCalled();
    });

    it('should fetch and sort giphies returned from API', () => {
      component = mountComponent();
      const instance = component.instance();

      instance.searchNode = {};
      instance.setState(() => ({
        searchVal: 'lebronto',
        sortDirection: 'desc',
      }));

      component.find('Pagination').props().onChange(instance.pageLimitOptions[0]);
      expect(props.actions.fetchGiphies).toHaveBeenCalled();

      props.actions.fetchGiphies()
        .then(() => {
          expect(props.actions.sortGiphies).toHaveBeenCalledWith('desc');
        });
    });

    it('should not perform search if there\'s no search value', () => {
      component = mountComponent();
      component.instance().searchNode = {};

      component.find('#search-btn').simulate('click');

      expect(props.actions.fetchGiphies).not.toHaveBeenCalled();
    });
  });
});
