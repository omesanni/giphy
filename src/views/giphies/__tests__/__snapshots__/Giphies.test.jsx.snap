// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Giphies renders 1`] = `
<section
  className="card card--emboss mt-5"
>
  <div
    className="card-header"
  >
    <div
      className="form-row"
    >
      <div
        className="form-group col-lg-4 col-sm-6 col-md-5 mb-0 mr-auto"
      >
        <div
          className="input-group"
        >
          <input
            className="form-control form-control-sm"
            id="search-input"
            placeholder="Search..."
            type="text"
          />
          <div
            className="input-group-append"
          >
            <button
              className="btn btn-outline-secondary"
              id="search-btn"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-search fa-w-16 "
                data-icon="search"
                data-prefix="fas"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="form-group col-lg-2 col-sm-3 justify-content-end d-flex mb-0"
      >
        <label
          className="mr-3 d-none d-sm-block text-muted"
        >
          Limit
        </label>
        <select
          className="form-control form-control-sm"
          defaultValue={25}
          id="page-limit-select"
          onChange={[Function]}
        >
          <option>
            10
          </option>
          <option>
            25
          </option>
          <option>
            50
          </option>
        </select>
      </div>
      <div
        className="form-group col-lg-1 col-sm-2 mb-0"
      >
        <select
          className="form-control form-control-sm"
          defaultValue="Order By"
          id="sort-select"
          onChange={[Function]}
        >
          <option
            disabled={true}
          >
            Order By
          </option>
          <option>
            asc
          </option>
          <option>
            desc
          </option>
        </select>
      </div>
    </div>
  </div>
  <div
    className="card-body"
  >
    <div
      className="position-relative"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-6 col-md-4 col-lg-3 mt-3"
        >
          <div
            className="card"
            id="giphy-aaa"
          >
            <img
              alt="cash me outside how bout dah"
              className="object-fit-cover cursor-pointer rounded-top"
              onClick={[Function]}
              src="http//www"
            />
            <div
              className="card-body card-body--rank-wrapper"
            >
              <div
                className="rank-controls"
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-arrow-up fa-w-14 rank-controls__up mr-3"
                  data-icon="arrow-up"
                  data-prefix="fas"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="badge badge-pill badge-secondary"
                >
                  20
                </span>
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-arrow-down fa-w-14 rank-controls__down ml-3"
                  data-icon="arrow-down"
                  data-prefix="fas"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col-sm-6 col-md-4 col-lg-3 mt-3"
        >
          <div
            className="card"
            id="giphy-bbb"
          >
            <img
              alt="put some respeck on my name"
              className="object-fit-cover cursor-pointer rounded-top"
              onClick={[Function]}
              src="http//www"
            />
            <div
              className="card-body card-body--rank-wrapper"
            >
              <div
                className="rank-controls"
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-arrow-up fa-w-14 rank-controls__up mr-3"
                  data-icon="arrow-up"
                  data-prefix="fas"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="badge badge-pill badge-secondary"
                >
                  30
                </span>
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-arrow-down fa-w-14 rank-controls__down ml-3"
                  data-icon="arrow-down"
                  data-prefix="fas"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col-sm-6 col-md-4 col-lg-3 mt-3"
        >
          <div
            className="card"
            id="giphy-ccc"
          >
            <img
              className="object-fit-cover cursor-pointer rounded-top"
              onClick={[Function]}
              src="http//www"
            />
            <div
              className="card-body card-body--rank-wrapper"
            >
              <div
                className="rank-controls"
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-arrow-up fa-w-14 rank-controls__up mr-3"
                  data-icon="arrow-up"
                  data-prefix="fas"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
                <span
                  className="badge badge-pill badge-secondary"
                >
                  10
                </span>
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-arrow-down fa-w-14 rank-controls__down ml-3"
                  data-icon="arrow-down"
                  data-prefix="fas"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <nav
        className="d-flex justify-content-center mt-5"
      >
        <ul
          className="pagination"
        >
          <li
            className="page-item disabled no-pointer-events"
            onClick={[Function]}
          >
            <span
              className="page-link"
            >
              Prev
            </span>
          </li>
          <li
            className="page-item active"
            onClick={[Function]}
          >
            <span
              className="page-link"
            >
              1
            </span>
          </li>
          <li
            className="page-item"
            onClick={[Function]}
          >
            <span
              className="page-link"
            >
              2
            </span>
          </li>
          <li
            className="page-item"
            onClick={[Function]}
          >
            <span
              className="page-link"
            >
              3
            </span>
          </li>
          <li
            className="page-item"
            onClick={[Function]}
          >
            <span
              className="page-link"
            >
              Next
            </span>
          </li>
        </ul>
      </nav>
    </div>
  </div>
  <div
    className="modal is-open"
  >
    <div
      className="modal__container"
    >
      <div
        className="modal__close"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          data-icon="times"
          data-prefix="fas"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
      <div
        className="modal__content"
      />
    </div>
    <div
      className="modal__backdrop"
      onClick={[Function]}
    />
  </div>
</section>
`;
